# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [ ]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

: 

# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Ticket_number,Ticket_item
0,1,0,3,Braund Mr Owen Harris,male,22.0,1,0,A/5 21171,7.2500,NaN,S,21171,A/5
1,2,1,1,Cumings Mrs John Bradley Florence Briggs Thayer,female,38.0,1,0,PC 17599,71.2833,C85,C,17599,PC
2,3,1,3,Heikkinen Miss Laina,female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,3101282,STON/O2.
3,4,1,1,Futrelle Mrs Jacques Heath Lily May Peel,female,35.0,1,0,113803,53.1000,C123,S,113803,NONE
4,5,0,3,Allen Mr William Henry,male,35.0,0,0,373450,8.0500,NaN,S,373450,NONE


Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2026-05-28T12:04:07.08188423+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4blskzek/model/ with prefix b45e584ced7f412e
[INFO 2026-05-28T12:04:07.092199917+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:04:07.092325216+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2026-05-28T12:04:10.587172205+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe71ysi5w/model/ with prefix f74a4b7c462c499a
[INFO 2026-05-28T12:04:10.596926297+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2026-05-28T12:04:10.597008682+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmp8umrub1e as temporary training directory


[INFO 2026-05-28T12:06:15.939225283+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8umrub1e/model/ with prefix 19d7fd1b0c014fa2
[INFO 2026-05-28T12:06:15.954477187+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2026-05-28T12:06:15.954539109+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2026-05-28T12:06:15.954577513+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2026-05-28T12:06:17.043400173+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq_qm_mwr/model/ with prefix 3ebf2f4c7e904cf2
[INFO 2026-05-28T12:06:17.04775734+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2026-05-28T12:06:18.538127859+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw_c3uw9o/model/ with prefix 3151c0bd1b06454c
[INFO 2026-05-28T12:06:18.557584175+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2026-05-28T12:06:19.656835544+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1u6hatir/model/ with prefix 1121446e806f45d1
[INFO 2026-05-28T12:06:19.661243265+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2026-05-28T12:06:21.35683681+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv3m3kcjp/model/ with prefix a7fc9ae47d534623
[INFO 2026-05-28T12:06:21.382227611+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2026-05-28T12:06:22.532915326+00:00 kernel.cc:1214] Loading model from path /tmp/tmpherhka1r/model/ with prefix 6d51fc060bed41fc
[INFO 2026-05-28T12:06:22.538675811+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2026-05-28T12:06:23.642816018+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuy035vey/model/ with prefix d51e7e9bbe584b8c
[INFO 2026-05-28T12:06:23.645913025+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2026-05-28T12:06:24.773929916+00:00 kernel.cc:1214] Loading model from path /tmp/tmptkm4_2rp/model/ with prefix 6b6a777ccebb495d
[INFO 2026-05-28T12:06:24.781293283+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2026-05-28T12:06:26.318937382+00:00 kernel.cc:1214] Loading model from path /tmp/tmppa0d3esk/model/ with prefix 803d01575fe24fd9
[INFO 2026-05-28T12:06:26.339411367+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:06:26.339468411+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2026-05-28T12:06:27.581260446+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkq8encd3/model/ with prefix 252677b3928e46b9
[INFO 2026-05-28T12:06:27.591239385+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2026-05-28T12:06:29.45152732+00:00 kernel.cc:1214] Loading model from path /tmp/tmp14kv0tls/model/ with prefix db8ae424c366405a
[INFO 2026-05-28T12:06:29.466899859+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2026-05-28T12:06:30.5654724+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxbvq86l6/model/ with prefix dbdf315d91e541d9
[INFO 2026-05-28T12:06:30.571407218+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2026-05-28T12:06:31.950358066+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9iae738a/model/ with prefix dc83efe39c9e4e06
[INFO 2026-05-28T12:06:31.965041636+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2026-05-28T12:06:33.062850519+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4ryex89l/model/ with prefix 504b632bcfaf47ef
[INFO 2026-05-28T12:06:33.068794935+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2026-05-28T12:06:34.32292932+00:00 kernel.cc:1214] Loading model from path /tmp/tmptxhwz3zy/model/ with prefix fee7d47757f54a4f
[INFO 2026-05-28T12:06:34.333845034+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2026-05-28T12:06:35.510158547+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfuij1m88/model/ with prefix ac6aa042f11f4188
[INFO 2026-05-28T12:06:35.516350892+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2026-05-28T12:06:36.630317229+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5az2eek0/model/ with prefix 6737d80c9a8a4260
[INFO 2026-05-28T12:06:36.637824092+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:06:36.637888633+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2026-05-28T12:06:37.926461189+00:00 kernel.cc:1214] Loading model from path /tmp/tmpaz3_2ms_/model/ with prefix fcbfc8a06a1446ac
[INFO 2026-05-28T12:06:37.938870185+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2026-05-28T12:06:39.219671743+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8g8tbke0/model/ with prefix d68284fb6ba04553
[INFO 2026-05-28T12:06:39.232728878+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2026-05-28T12:06:40.479863407+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu56hphfz/model/ with prefix c241a54212d74298
[INFO 2026-05-28T12:06:40.491227334+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2026-05-28T12:06:41.990050954+00:00 kernel.cc:1214] Loading model from path /tmp/tmpt89hur8r/model/ with prefix 5f802ed49d914928
[INFO 2026-05-28T12:06:42.009312232+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2026-05-28T12:06:43.376316801+00:00 kernel.cc:1214] Loading model from path /tmp/tmp602fy_55/model/ with prefix 04798026e3054c99
[INFO 2026-05-28T12:06:43.39110072+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2026-05-28T12:06:44.469001005+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzjtwdzjr/model/ with prefix e86984f33cb14815
[INFO 2026-05-28T12:06:44.474274147+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2026-05-28T12:06:45.542037927+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj8tqw7g1/model/ with prefix 748fbe5f4d624624
[INFO 2026-05-28T12:06:45.547923683+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2026-05-28T12:06:46.676029906+00:00 kernel.cc:1214] Loading model from path /tmp/tmpikd8gd4j/model/ with prefix 73fac04ae6334680
[INFO 2026-05-28T12:06:46.684987026+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:06:46.685034101+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2026-05-28T12:06:47.767452997+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjnga5irt/model/ with prefix ce90825d26f04a4b
[INFO 2026-05-28T12:06:47.773334029+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2026-05-28T12:06:49.010867944+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe2zai7ye/model/ with prefix 41f0003738044f19
[INFO 2026-05-28T12:06:49.022120387+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2026-05-28T12:06:50.229696771+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv3wqsnhe/model/ with prefix 6588166b1e064e13
[INFO 2026-05-28T12:06:50.239636971+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2026-05-28T12:06:51.364830333+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcs095d8n/model/ with prefix 8c3d05ede6ad4c7a
[INFO 2026-05-28T12:06:51.371424161+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2026-05-28T12:06:52.495919934+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd7e9vmrq/model/ with prefix aeb5a82d75f94c42
[INFO 2026-05-28T12:06:52.501400766+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2026-05-28T12:06:53.783465737+00:00 kernel.cc:1214] Loading model from path /tmp/tmplm6coeyb/model/ with prefix ccf471131d004f5f
[INFO 2026-05-28T12:06:53.796247718+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2026-05-28T12:06:55.523013724+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwwdfjler/model/ with prefix e9d7b984946a4cdf
[INFO 2026-05-28T12:06:55.549450541+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2026-05-28T12:06:57.394823326+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf_yfz2jt/model/ with prefix 822180b974814cd2
[INFO 2026-05-28T12:06:57.406911995+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:06:57.406960874+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2026-05-28T12:06:58.590196428+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6h7alvo2/model/ with prefix b8fec79993a341ad
[INFO 2026-05-28T12:06:58.596706319+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2026-05-28T12:06:59.973314145+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6cugydje/model/ with prefix 252b47a0c9ef4580
[INFO 2026-05-28T12:06:59.986865723+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2026-05-28T12:07:01.229300997+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnj914dkt/model/ with prefix e4a1cd7d139044ed
[INFO 2026-05-28T12:07:01.23836224+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2026-05-28T12:07:02.570707706+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0zo89bme/model/ with prefix cb1a36abf6134581
[INFO 2026-05-28T12:07:02.578452864+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2026-05-28T12:07:03.942891582+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbjowu8p0/model/ with prefix 65541f69c48d4a8a
[INFO 2026-05-28T12:07:03.957079548+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2026-05-28T12:07:05.167823441+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_f4m55de/model/ with prefix ba653472eb8a47eb
[INFO 2026-05-28T12:07:05.177049166+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2026-05-28T12:07:06.63013386+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqq7ghngv/model/ with prefix 56e49670e11c4b9b
[INFO 2026-05-28T12:07:06.643521232+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2026-05-28T12:07:08.027831749+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq38y8w8i/model/ with prefix b72ceafa9ade40a8
[INFO 2026-05-28T12:07:08.040512308+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:07:08.04056999+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2026-05-28T12:07:09.113082609+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbp8l8bjl/model/ with prefix d971862a7ca3482d
[INFO 2026-05-28T12:07:09.117723646+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2026-05-28T12:07:10.521594461+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk5xzluyd/model/ with prefix 8c06e2317fa6413a
[INFO 2026-05-28T12:07:10.537159738+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2026-05-28T12:07:11.778252344+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjd_sdvkk/model/ with prefix f759b690ff7b4364
[INFO 2026-05-28T12:07:11.78930271+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2026-05-28T12:07:13.321727158+00:00 kernel.cc:1214] Loading model from path /tmp/tmphq_054l2/model/ with prefix 38ca38c54f5d4738
[INFO 2026-05-28T12:07:13.338559842+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2026-05-28T12:07:14.567569867+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnym8dlt8/model/ with prefix 66bc131c598647bb
[INFO 2026-05-28T12:07:14.576935227+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2026-05-28T12:07:15.59812717+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjdbsvgwl/model/ with prefix c7a5e9fbf11441c5
[INFO 2026-05-28T12:07:15.602075131+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2026-05-28T12:07:16.971898567+00:00 kernel.cc:1214] Loading model from path /tmp/tmp18j72evv/model/ with prefix 2ab9ff95f9574c5b
[INFO 2026-05-28T12:07:16.986807281+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2026-05-28T12:07:18.341116874+00:00 kernel.cc:1214] Loading model from path /tmp/tmpscrbqa6a/model/ with prefix 12425454ac0941ab
[INFO 2026-05-28T12:07:18.35426885+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:07:18.354315062+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2026-05-28T12:07:19.419332478+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy0ef4zpg/model/ with prefix 2f3dbc08349c44f2
[INFO 2026-05-28T12:07:19.423849272+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2026-05-28T12:07:20.547655685+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsx2g9fnb/model/ with prefix 3287bb666bb84e52
[INFO 2026-05-28T12:07:20.554450877+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2026-05-28T12:07:21.820047625+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1t5wxibk/model/ with prefix f9f10f5fbc234f19
[INFO 2026-05-28T12:07:21.831918198+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2026-05-28T12:07:23.225351709+00:00 kernel.cc:1214] Loading model from path /tmp/tmp89n92vxo/model/ with prefix 195f59bc67e04282
[INFO 2026-05-28T12:07:23.240637983+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2026-05-28T12:07:24.389681905+00:00 kernel.cc:1214] Loading model from path /tmp/tmp24hqlkwp/model/ with prefix d84863ca8a4244e4
[INFO 2026-05-28T12:07:24.397206148+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2026-05-28T12:07:25.536623192+00:00 kernel.cc:1214] Loading model from path /tmp/tmph3amcvx9/model/ with prefix be2979840fe94bf9
[INFO 2026-05-28T12:07:25.54392049+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2026-05-28T12:07:26.580466243+00:00 kernel.cc:1214] Loading model from path /tmp/tmph_jp0fze/model/ with prefix f74e2627911a486e
[INFO 2026-05-28T12:07:26.583815551+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2026-05-28T12:07:27.924831042+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_vh3z1o6/model/ with prefix bcf5d47a52e442a1
[INFO 2026-05-28T12:07:27.939180495+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2026-05-28T12:07:29.865762514+00:00 kernel.cc:1214] Loading model from path /tmp/tmpboe3s693/model/ with prefix 20712155e4234386
[INFO 2026-05-28T12:07:29.879033445+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:07:29.879086677+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2026-05-28T12:07:31.088383576+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4ho_26ki/model/ with prefix 1556527c418c4cf7
[INFO 2026-05-28T12:07:31.093853346+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2026-05-28T12:07:32.424110569+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnr5fh83h/model/ with prefix 493079ffc1d74502
[INFO 2026-05-28T12:07:32.432530037+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2026-05-28T12:07:33.758930871+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9q72dpn6/model/ with prefix 4a72d4c61c914d3d
[INFO 2026-05-28T12:07:33.768348128+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2026-05-28T12:07:35.064174792+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5gw9keh3/model/ with prefix 5df2064c358a4c28
[INFO 2026-05-28T12:07:35.076593559+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2026-05-28T12:07:36.233294662+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyeg7_pqo/model/ with prefix 0d0d6854321d493b
[INFO 2026-05-28T12:07:36.238247957+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2026-05-28T12:07:37.978656142+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnt6tjxgr/model/ with prefix 81a8ccec910348d2
[INFO 2026-05-28T12:07:38.004883326+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2026-05-28T12:07:39.202967129+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6_c6wgbo/model/ with prefix c0eee20da1ad44b3
[INFO 2026-05-28T12:07:39.212426183+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2026-05-28T12:07:40.403880578+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq5b31cx3/model/ with prefix b420c992bdb14709
[INFO 2026-05-28T12:07:40.411973099+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:07:40.412017828+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2026-05-28T12:07:41.504085101+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxgo1az3_/model/ with prefix 37a3cb66f61341b6
[INFO 2026-05-28T12:07:41.509359603+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2026-05-28T12:07:42.660998649+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbrkzk1j9/model/ with prefix 6f22a45fe90c439b
[INFO 2026-05-28T12:07:42.667177188+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2026-05-28T12:07:44.134286208+00:00 kernel.cc:1214] Loading model from path /tmp/tmplsga0nkl/model/ with prefix f333db0183a8481b
[INFO 2026-05-28T12:07:44.152403357+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2026-05-28T12:07:45.413057996+00:00 kernel.cc:1214] Loading model from path /tmp/tmprsmkhrd2/model/ with prefix a56153babd004857
[INFO 2026-05-28T12:07:45.424612881+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2026-05-28T12:07:46.560486716+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9u23d8pk/model/ with prefix 201be37acd854d50
[INFO 2026-05-28T12:07:46.567603214+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2026-05-28T12:07:47.778571056+00:00 kernel.cc:1214] Loading model from path /tmp/tmptafclv0k/model/ with prefix 3ec1b181cbb94337
[INFO 2026-05-28T12:07:47.78733898+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2026-05-28T12:07:48.945344287+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6dg_k_q7/model/ with prefix 49ae8985f74e41bc
[INFO 2026-05-28T12:07:48.953651314+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2026-05-28T12:07:50.340059191+00:00 kernel.cc:1214] Loading model from path /tmp/tmppvq_dj5n/model/ with prefix cf5554ac838e41d2
[INFO 2026-05-28T12:07:50.355578401+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2026-05-28T12:07:51.525630987+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0e4b3t4i/model/ with prefix 4d52e4a723be4443
[INFO 2026-05-28T12:07:51.535834269+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:07:51.535902531+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2026-05-28T12:07:52.813116503+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsw1syeyp/model/ with prefix 5d71ec8780a3429d
[INFO 2026-05-28T12:07:52.824629028+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2026-05-28T12:07:54.01175427+00:00 kernel.cc:1214] Loading model from path /tmp/tmpi904794_/model/ with prefix 81a5b15a5ebb4de4
[INFO 2026-05-28T12:07:54.019715047+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2026-05-28T12:07:55.123391403+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfab4vto8/model/ with prefix 7593af06a2c146b8
[INFO 2026-05-28T12:07:55.127544551+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2026-05-28T12:07:56.193325202+00:00 kernel.cc:1214] Loading model from path /tmp/tmpanvybawt/model/ with prefix b491bb2dbadc45fb
[INFO 2026-05-28T12:07:56.198215602+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2026-05-28T12:07:57.32801269+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjo3kpzx1/model/ with prefix 6a804ce11309452c
[INFO 2026-05-28T12:07:57.334592184+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2026-05-28T12:07:58.486423166+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwqtnj2r4/model/ with prefix c4714fdfc6b8427a
[INFO 2026-05-28T12:07:58.493302056+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2026-05-28T12:07:59.697733298+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkd7_f3a2/model/ with prefix 394cd47cb53340bf
[INFO 2026-05-28T12:07:59.707801692+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2026-05-28T12:08:00.997258793+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj46kbzt0/model/ with prefix 839a64563d6f4367
[INFO 2026-05-28T12:08:01.008161261+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2026-05-28T12:08:02.261677258+00:00 kernel.cc:1214] Loading model from path /tmp/tmp24l_onhy/model/ with prefix a2920f51de1a4ef4
[INFO 2026-05-28T12:08:02.271778331+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:08:02.271825878+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2026-05-28T12:08:03.4573051+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjqpmal_y/model/ with prefix b34c049d44864a56
[INFO 2026-05-28T12:08:03.465895628+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2026-05-28T12:08:05.686836355+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_fil4j2f/model/ with prefix 7f9500a0f0724cd2
[INFO 2026-05-28T12:08:05.705806421+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2026-05-28T12:08:06.899659233+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfxq091pv/model/ with prefix feb3e804d7f94b86
[INFO 2026-05-28T12:08:06.907063093+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2026-05-28T12:08:08.356451841+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq3zf6xsx/model/ with prefix 711987b4511946c5
[INFO 2026-05-28T12:08:08.372904593+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2026-05-28T12:08:09.835002827+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7hu9sue1/model/ with prefix b28cf3205d884dc3
[INFO 2026-05-28T12:08:09.853795972+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2026-05-28T12:08:11.125953103+00:00 kernel.cc:1214] Loading model from path /tmp/tmpax1anpnx/model/ with prefix 822f59171e9f4549
[INFO 2026-05-28T12:08:11.137251834+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2026-05-28T12:08:12.269082018+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvkr01exz/model/ with prefix faa34a1c4c30442f
[INFO 2026-05-28T12:08:12.274064021+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:08:12.27410725+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2026-05-28T12:08:13.502175868+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfthkmbyc/model/ with prefix c4ee013b866a48d0
[INFO 2026-05-28T12:08:13.512599265+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2026-05-28T12:08:14.675566277+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7eqc_5e9/model/ with prefix f659669d60b54beb
[INFO 2026-05-28T12:08:14.683000412+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2026-05-28T12:08:16.104842695+00:00 kernel.cc:1214] Loading model from path /tmp/tmpalsw27c2/model/ with prefix f994474156364c75
[INFO 2026-05-28T12:08:16.122102091+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2026-05-28T12:08:17.390881739+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9ld8mufw/model/ with prefix 5cd7b3586a964f69
[INFO 2026-05-28T12:08:17.400771435+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2026-05-28T12:08:18.526948811+00:00 kernel.cc:1214] Loading model from path /tmp/tmpn3x9ay3l/model/ with prefix 8c405282f0a04f16
[INFO 2026-05-28T12:08:18.533631209+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2026-05-28T12:08:19.743433675+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7mr7gf9y/model/ with prefix a6dedf81e1ed4378
[INFO 2026-05-28T12:08:19.751873915+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2026-05-28T12:08:21.159096357+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsdzq0g85/model/ with prefix 9f306cc3963140e8
[INFO 2026-05-28T12:08:21.171547109+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2026-05-28T12:08:22.374229794+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2wf_td2s/model/ with prefix 9807c5ec12f44d95
[INFO 2026-05-28T12:08:22.379969786+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-05-28T12:08:22.380025194+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2026-05-28T12:08:23.620744156+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7k4py5fx/model/ with prefix a7b7c4afcc3c41ac
[INFO 2026-05-28T12:08:23.628429459+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2026-05-28T12:08:25.107966613+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdjwjztzk/model/ with prefix 7b381a8b9d844696
[INFO 2026-05-28T12:08:25.122030522+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 